j.u.ConcurrentModificationException with Scala 2.13.{4,5} and Java sources

We get a `java.util.ConcurrentModificationException` when compiling code with zinc and Scala 2.13.4 or 2.13.5. The same code can be compiled successfully, when Scala 2.13.3 or older is used.

## steps

1. Clone the minimal reproducer project from: https://github.com/chikei/zinc-ConcurrentModificationException (https://github.com/chikei/zinc-ConcurrentModificationException/commit/96b0e904481cf82da8557ea7d18dfbc3302c546e)

2. run `sbt run`

## problem

Instead of compiling successfully, an `java.util.ConcurrentModificationException` is thrown.

<details>
<summary>The full output:</summary>

```
21:39:28 ⌁ θ62° lefou:~/work/tmp/zinc-ConcurrentModificationException master ± sbt run
[info] welcome to sbt 1.4.7 (AdoptOpenJDK Java 1.8.0_272)
[info] loading settings for project global-plugins from coursier.sbt,eclipse.sbt,idea.sbt,metals.sbt ...
[info] loading global plugins from /home/lefou/.sbt/1.0/plugins
[info] loading project definition from /home/lefou/work/tmp/zinc-ConcurrentModificationException/project
[info] loading settings for project zinc-concurrentmodificationexception from build.sbt ...
[info] set current project to zinc-concurrent (in build file:/home/lefou/work/tmp/zinc-ConcurrentModificationException/)
[info] compiling 1 Scala source to /home/lefou/work/tmp/zinc-ConcurrentModificationException/target/scala-2.13/classes ...
[warn] 2 deprecations (since 1.4.0); re-run with -deprecation for details
[warn] one warning found
[info] running example.MainClass 
[info] compiling 1 Java source to /tmp/zincClasses3378735785364370019 ...
[error] ## Exception when compiling 1 sources to /tmp/zincClasses3378735785364370019
[error] java.util.ConcurrentModificationException: mutation occurred during iteration
[error] scala.collection.mutable.MutationTracker$.checkMutations(MutationTracker.scala:43)
[error] scala.collection.mutable.MutationTracker$CheckedIterator.hasNext(MutationTracker.scala:59)
[error] scala.collection.IterableOnceOps.foreach(IterableOnce.scala:563)
[error] scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:561)
[error] scala.collection.AbstractIterable.foreach(Iterable.scala:919)
[error] sbt.internal.inc.ClassToAPI$.process(ClassToAPI.scala:34)
[error] sbt.internal.inc.javac.AnalyzingJavaCompiler.readAPI$1(AnalyzingJavaCompiler.scala:180)
[error] sbt.internal.inc.javac.AnalyzingJavaCompiler.$anonfun$compile$23(AnalyzingJavaCompiler.scala:205)
[error] sbt.internal.inc.classfile.JavaAnalyze$.readInheritanceDependencies$1(JavaAnalyze.scala:194)
[error] sbt.internal.inc.classfile.JavaAnalyze$.$anonfun$apply$15(JavaAnalyze.scala:201)
[error] sbt.internal.inc.classfile.JavaAnalyze$.$anonfun$apply$15$adapted(JavaAnalyze.scala:110)
[error] scala.collection.IterableOnceOps.foreach(IterableOnce.scala:563)
[error] scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:561)
[error] scala.collection.AbstractIterable.foreach(Iterable.scala:919)
[error] scala.collection.IterableOps$WithFilter.foreach(Iterable.scala:889)
[error] sbt.internal.inc.classfile.JavaAnalyze$.apply(JavaAnalyze.scala:110)
[error] sbt.internal.inc.javac.AnalyzingJavaCompiler.$anonfun$compile$21(AnalyzingJavaCompiler.scala:205)
[error] sbt.internal.inc.javac.AnalyzingJavaCompiler.$anonfun$compile$21$adapted(AnalyzingJavaCompiler.scala:199)
[error] scala.collection.IterableOnceOps.foreach(IterableOnce.scala:563)
[error] scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:561)
[error] scala.collection.AbstractIterable.foreach(Iterable.scala:919)
[error] scala.collection.IterableOps$WithFilter.foreach(Iterable.scala:889)
[error] sbt.internal.inc.javac.AnalyzingJavaCompiler.$anonfun$compile$19(AnalyzingJavaCompiler.scala:199)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[error] sbt.internal.inc.javac.AnalyzingJavaCompiler.timed(AnalyzingJavaCompiler.scala:251)
[error] sbt.internal.inc.javac.AnalyzingJavaCompiler.compile(AnalyzingJavaCompiler.scala:199)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compileJava$1(MixedAnalyzingCompiler.scala:98)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[error] sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:241)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compileJava(MixedAnalyzingCompiler.scala:61)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compileJava0$1(MixedAnalyzingCompiler.scala:191)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:204)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:528)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:528)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:174)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:172)
[error] sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:457)
[error] sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
[error] sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:261)
[error] sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:412)
[error] sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:499)
[error] sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:399)
[error] sbt.internal.inc.Incremental$.apply(Incremental.scala:166)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:528)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:482)
[error] sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:420)
[error] sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
[error] example.MainClass$.delayedEndpoint$example$MainClass$1(MainClass.scala:85)
[error] example.MainClass$delayedInit$body.apply(MainClass.scala:13)
[error] scala.Function0.apply$mcV$sp(Function0.scala:39)
[error] scala.Function0.apply$mcV$sp$(Function0.scala:39)
[error] scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:17)
[error] scala.App.$anonfun$main$1(App.scala:76)
[error] scala.App.$anonfun$main$1$adapted(App.scala:76)
[error] scala.collection.IterableOnceOps.foreach(IterableOnce.scala:563)
[error] scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:561)
[error] scala.collection.AbstractIterable.foreach(Iterable.scala:919)
[error] scala.App.main(App.scala:76)
[error] scala.App.main$(App.scala:74)
[error] example.MainClass$.main(MainClass.scala:13)
[error] example.MainClass.main(MainClass.scala)
[error] sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[error] sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[error] sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[error] java.lang.reflect.Method.invoke(Method.java:498)
[error] sbt.Run.invokeMain(Run.scala:133)
[error] sbt.Run.execute$1(Run.scala:82)
[error] sbt.Run.$anonfun$runWithLoader$5(Run.scala:110)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] sbt.util.InterfaceUtil$$anon$1.get(InterfaceUtil.scala:17)
[error] sbt.TrapExit$App.run(TrapExit.scala:258)
[error] java.lang.Thread.run(Thread.java:748)
[error]            
[error] (run-main-0) java.util.ConcurrentModificationException: mutation occurred during iteration
[error] java.util.ConcurrentModificationException: mutation occurred during iteration
[error]         at scala.collection.mutable.MutationTracker$.checkMutations(MutationTracker.scala:43)
[error]         at scala.collection.mutable.MutationTracker$CheckedIterator.hasNext(MutationTracker.scala:59)
[error]         at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:563)
[error]         at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:561)
[error]         at scala.collection.AbstractIterable.foreach(Iterable.scala:919)
[error] stack trace is suppressed; run last Compile / bgRun for the full output
[error] Nonzero exit code: 1
[error] (Compile / run) Nonzero exit code: 1
[error] Total time: 8 s, completed 15.04.2021 21:42:15
```
</details>

## expectation

The compilation should succeed successfully.

## notes

We hit this issue in mill (https://github.com/com-lihaoyi/mill), which uses zinc to compile Scala and Java projects. This issue is a blocker to update to Scala 2.13.5, as we can't compile Mills codebase. See also discussion https://github.com/com-lihaoyi/mill/discussions/1278

The reproducer uses zinc 1.5.0 and fails with AdoptOpenJDK 1.8 as well as AdoptOpenJDK 11 on Linux.

It is a cut down version of our use of zinc in Mill (https://github.com/com-lihaoyi/mill/blob/master/scalalib/worker/src/ZincWorkerImpl.scala).

The repro was created by @chikei and has the following structure:

```
.
├── README.md
├── build.sbt
├── project
│   └── build.properties
└── src
    └── main
        ├── resources
        │   └── mill
        │       └── main
        │           └── client
        │               └── FileToStreamTailer.java  <-- The Java source file to compile
        └── scala
            └── example
                └── MainClass.scala  <-- A minimal App using the zinc compiler to compile the sources under resources

```


Here is a GH Actions workflow that shows the failing CI for mill: https://github.com/com-lihaoyi/mill/actions/runs/746276787

Ah, you mean latest stable sbt? The repro uses sbt 1.4.7. After bumping to 1.5.0 I get the following `LinkageErrror`:

```
07:20:06 ⌁ θ62° lefou:~/work/tmp/zinc-ConcurrentModificationException master(+1/-1) 1 ± sbt run
[info] welcome to sbt 1.5.0 (AdoptOpenJDK Java 1.8.0_272)
[info] loading settings for project global-plugins from coursier.sbt,eclipse.sbt,idea.sbt,metals.sbt ...
[info] loading global plugins from /home/lefou/.sbt/1.0/plugins
[info] loading project definition from /home/lefou/work/tmp/zinc-ConcurrentModificationException/project
[info] loading settings for project zinc-concurrentmodificationexception from build.sbt ...
[info] set current project to zinc-concurrent (in build file:/home/lefou/work/tmp/zinc-ConcurrentModificationException/)
[info] running example.MainClass 
[error] (run-main-0) java.lang.BootstrapMethodError: java.lang.LinkageError: loader constraint violation: loader (instance of sbt/internal/LayeredClassLoader) previously initiated loading for a different type with name "xsbti/Position"
[error] java.lang.BootstrapMethodError: java.lang.LinkageError: loader constraint violation: loader (instance of sbt/internal/LayeredClassLoader) previously initiated loading for a different type with name "xsbti/Position"
[error] Caused by: java.lang.LinkageError: loader constraint violation: loader (instance of sbt/internal/LayeredClassLoader) previously initiated loading for a different type with name "xsbti/Position"
[error]         at java.lang.ClassLoader.defineClass1(Native Method)
[error]         at java.lang.ClassLoader.defineClass(ClassLoader.java:756)
[error]         at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
[error]         at java.net.URLClassLoader.defineClass(URLClassLoader.java:468)
[error]         at java.net.URLClassLoader.access$100(URLClassLoader.java:74)
[error]         at java.net.URLClassLoader$1.run(URLClassLoader.java:369)
[error]         at java.net.URLClassLoader$1.run(URLClassLoader.java:363)
[error]         at java.security.AccessController.doPrivileged(Native Method)
[error]         at java.net.URLClassLoader.findClass(URLClassLoader.java:362)
[error] stack trace is suppressed; run last Compile / bgRun for the full output
[error] Nonzero exit code: 1
[error] (Compile / run) Nonzero exit code: 1
[error] Total time: 2 s, completed 16.04.2021 07:20:18
```

I think we have a conflict here between sbt itself and the used zinc in the main. The real issue is unrelated to the executing build tool, though. @chikei was also able to reproduce it in a ammonite session, without any sbt (except the zinc) involved. https://gist.github.com/chikei/413956ffbf737402a3ae96f2010635c7

--- a/internal/zinc-apiinfo/src/main/scala/sbt/internal/inc/ClassToAPI.scala
+++ b/internal/zinc-apiinfo/src/main/scala/sbt/internal/inc/ClassToAPI.scala
@@ -31,7 +31,8 @@ object ClassToAPI {
   ): (Seq[api.ClassLike], Seq[String], Set[(Class[_], Class[_])]) = {
     val cmap = emptyClassMap
     classes.foreach(toDefinitions(cmap)) // force recording of class definitions
-    cmap.lz.foreach(_.get()) // force thunks to ensure all inherited dependencies are recorded
+    cmap.lz.toList
+      .foreach(_.get()) // force thunks to ensure all inherited dependencies are recorded
     val classApis = cmap.allNonLocalClasses.toSeq
     val mainClasses = cmap.mainClasses.toSeq
     val inDeps = cmap.inherited.toSet

